cunumeric/module.py

```diff
@@ -6227,3 +6227,740 @@ def bincount(
+    reshuffled_axes = tuple(range(min_dim_index, min_dim_index + num_axes))
+
+    non_consecutive = is_diff(sorted_axes, reshuffled_axes)
+    if non_consecutive:
+        arr_shuffled = moveaxis(arr, sorted_axes, reshuffled_axes)
+    else:
+        arr_shuffled = arr
+
+    # shape_reshuffled = arr_shuffled.shape # debug
+    collapsed_shape = np.product(
+        [arr_shuffled.shape[i] for i in reshuffled_axes]
+    )
+
+    redimed = tuple(range(0, min_dim_index + 1)) + tuple(
+        range(min_dim_index + num_axes, ndim)
+    )
+    reshaped = tuple(
+        [
+            collapsed_shape if k == min_dim_index else arr_shuffled.shape[k]
+            for k in redimed
+        ]
+    )
+
+    arr_reshaped = arr_shuffled.reshape(reshaped)
+    return (min_dim_index, arr_reshaped)
+
+
+# args:
+#
+# arr:      [in] source nd-array on which quantiles are calculated;
+#                preccondition: assumed sorted!
+# q_arr:    [in] quantile input values nd-array;
+# axis:     [in] axis along which quantiles are calculated;
+# method:   [in] func(q, n) returning (gamma, j),
+#                where = array1D.size;
+# keepdims: [in] boolean flag specifying whether collapsed axis
+#                should be kept as dim=1;
+# to_dtype: [in] dtype to convert the result to;
+# qs_all:   [in/out] result pass through or created (returned)
+# input_is_scalar: [in] specifies if q input was scalar
+# return: nd-array of quantile output values
+#         where its shape is obtained as:
+#         concatenating q_arr.shape with arr.shape \ {axis}
+#         (the shape of `arr` obtained by taking the `axis` out)
+#
+def quantile_impl(
+    arr,
+    q_arr,
+    axis,
+    axes_set,
+    original_shape,
+    method,
+    keepdims,
+    to_dtype,
+    qs_all,
+    input_is_scalar,
+):
+
+    ndims = len(arr.shape)
+
+    if axis is None:
+        n = arr.size
+        remaining_shape = []  # only `q_arr` dictates shape;
+        # quantile applied to `arr` seen as 1D;
+    else:
+        n = arr.shape[axis]
+
+        # arr.shape -{axis}; if keepdims use 1 for arr.shape[axis]:
+        # (can be empty [])
+        #
+        if keepdims:
+            remaining_shape = [
+                1 if k in axes_set else original_shape[k]
+                for k in range(0, len(original_shape))
+            ]
+        else:
+            remaining_shape = [
+                arr.shape[k] for k in range(0, ndims) if k != axis
+            ]
+
+    # quantile interpolation method to act on vectors,
+    # like an axpy call:
+    # [gamma](vleft, vright) { return (1-gamma)*vleft + gamma*vright;}
+    #
+    linear_interp = (
+        lambda gamma, arr_lvals, arr_rvals: (1.0 - gamma) * arr_lvals
+        + gamma * arr_rvals
+    )
+
+    # flattening to 1D might be the right way,
+    # but may consider profiling other approaches
+    # (slice assignment)
+    #
+    q_flat = q_arr.flatten().tolist()
+
+    # TODO: check if a container more efficient than list
```

Comment:
I think it will be more efficient to pre-allocate an array of the final shape that you want to return (or use `qs_all` directly, if it has been provided). Then you can fill in slices of this array like so:
```
for index, q in np.ndenumerate(q_arr):
  ...
  quantiles_ls = ...
  qs_all[index] = quantiles_ls
```

Reply:
This approach fails for scalar `q`. In which case `qs_array = np.asarray(q)` has an empty tuple as index:
```
>>> qsc = np.asarray(0.5)
>>> for index, q in np.ndenumerate(qsc):
...  print(index, q)
... 
() 0.5
>>>
```
I'm trying to figure out a way to enclose this case wo/ having to bifurcate the code into a special case for `isscalar(q)` (which I've been struggling to avoid from the get-go)...

Reply:
The RHS, `quantiles_ls` is always a 1D array. So, the problem is the shapes of LHS and RHS are not the same. The total size is the same, but not the same. Example:
```
err.txt:15361:lhs[index] shape=(1, 3, 1); rhs shape = (3,)
err.txt:15462:lhs[index] shape=(2, 1); rhs shape = (2,)

err.txt:16776:lhs[index] shape=(8, 1); rhs shape = (8,)
err.txt:18150:lhs[:] shape=(3, 4); rhs shape = (12,)
```
which is why the copy LHS[...] = RHS fails.

Reply:
Found a compromise: reshape() the RHS, for which failing cases differed only by an additional size 1 dimension (hence equivalent size). I assume reshaping in this case would not take a large toll (if any). Also RHS is always (typically _much_) smaller than the overall result.

Reply:
Yes, `reshape()` should be a noop if it's just adding or removing singleton dimensions.